Sources/NIO/ByteBuffer-aux.swift

```diff
@@ -431,4 +431,283 @@ extension ByteBuffer {
+        return buffer
+    }
+
+    /// Create a fresh `ByteBuffer` containing the readable bytes of `buffer`.
+    ///
+    /// This may allocate a new `ByteBuffer` with enough space fitting `buffer` and potentially some extra space.
+    ///
+    /// - note: Use this method only if you deliberately want to reallocate a potentially smaller `ByteBuffer` than the
+    ///         one you already have. Given that `ByteBuffer` is a value type, defensive copies are not necessary. If
+    ///         you have a `ByteBuffer` but would like the `readerIndex` to start at `0`, consider `readSlice` instead.
+    ///
+    /// - returns: The `ByteBuffer` containing the written bytes.
+    public func buffer(buffer: ByteBuffer) -> ByteBuffer {
+        var newBuffer = self.buffer(capacity: buffer.readableBytes)
+        newBuffer.writeImmutableBuffer(buffer)
+        return newBuffer
+    }
+
+    /// Create a fresh `ByteBuffer` containing the bytes contained in the given `DispatchData`.
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space fitting the bytes of the `DispatchData` and potentially
+    /// some extra space.
+    ///
+    /// - returns: The `ByteBuffer` containing the written bytes.
+    public func buffer(dispatchData: DispatchData) -> ByteBuffer {
+        var buffer = self.buffer(capacity: dispatchData.count)
+        buffer.writeDispatchData(dispatchData)
+        return buffer
+    }
+}
+
+
+extension Optional where Wrapped == ByteBuffer {
+    /// If `nil`, replace `self` with `.some(buffer)`. If non-`nil`, write `buffer`'s readable bytes into the
+    /// `ByteBuffer` starting at `writerIndex`.
+    ///
+    ///  This method will not modify `buffer`, meaning its `readerIndex` and `writerIndex` stays intact.
+    ///
+    /// - parameters:
+    ///     - buffer: The `ByteBuffer` to write.
+    /// - returns: The number of bytes written to this `ByteBuffer` which is equal to the number of `readableBytes` in
+    ///            buffer`.
+    @discardableResult
+    public mutating func writeImmutableBuffer(_ buffer: ByteBuffer) -> Int {
+        var mutable = buffer
+        return self.writeBuffer(&mutable)
+    }
+
+    /// If `nil`, replace `self` with `.some(buffer)`. If non-`nil`, write `buffer`'s readable bytes into the
+    /// `ByteBuffer` starting at `writerIndex`.
+    ///
+    /// This will move both this `ByteBuffer`'s writer index as well as `buffer`'s reader index by the number of bytes
+    /// readable in `buffer`.
+    ///
+    /// - parameters:
+    ///     - buffer: The `ByteBuffer` to write.
+    /// - returns: The number of bytes written to this `ByteBuffer` which is equal to the number of bytes read from `buffer`.
+    @discardableResult
+    public mutating func writeBuffer(_ buffer: inout ByteBuffer) -> Int {
+        if self == nil {
+            let input = buffer
+            self = input
+            buffer.moveReaderIndex(to: buffer.writerIndex)
+            return buffer.readableBytes
```

Comment:
This return value is wrong. `buffer.readableBytes` will be zero here, always, because we just moved the reader index.

Reply:
Great catch, thank you!!

I also renamed those methods to `setOrWriteBuffer`. The problem with the old spelling was that if you type

```
self.optionalBB.writeBuffer
```

then Xcode suggests to replace this with

```
self.optionalBB?.writeBuffer
```

ie, optional chaining which would lose the write!! Now, I chose a method name for the optional bit that doesn't exist on BB so this can't happen.

Reply:
Yeah, happy with those names.